// ts-json-schema-generator/dist/src/ChainNodeParser.js
import { UnknownNodeError } from "./Error/UnknownNodeError.js";
import { ReferenceType } from "./Type/ReferenceType.js";
export class ChainNodeParser {
    typeChecker;
    nodeParsers;
    typeCaches = new WeakMap();
    constructor(typeChecker, nodeParsers) {
        this.typeChecker = typeChecker;
        this.nodeParsers = nodeParsers;
    }
    addNodeParser(nodeParser) {
        this.nodeParsers.push(nodeParser);
        return this;
    }
    supportsNode(node) {
        return this.nodeParsers.some((nodeParser) => nodeParser.supportsNode(node));
    }
    createType(node, context, reference) {
        let typeCache = this.typeCaches.get(node);
        if (typeCache == null) {
            typeCache = new Map();
            this.typeCaches.set(node, typeCache);
        }
        const contextCacheKey = context.getCacheKey();
        let type = typeCache.get(contextCacheKey);
        if (!type) {
            type = this.getNodeParser(node, context).createType(node, context, reference);
            if (!(type instanceof ReferenceType)) {
                typeCache.set(contextCacheKey, type);
            }
        }
        return type;
    }
    getNodeParser(node, context) {
        for (const nodeParser of this.nodeParsers) {
            if (nodeParser.supportsNode(node)) {
                return nodeParser;
            }
        }
        throw new UnknownNodeError(node, context.getReference());
    }
}
