// ts-json-schema-generator/dist/src/ExposeNodeParser.js
import ts from "typescript";
import { DefinitionType } from "./Type/DefinitionType.js";
import { hasJsDocTag } from "./Utils/hasJsDocTag.js";
import { symbolAtNode } from "./Utils/symbolAtNode.js";
export class ExposeNodeParser {
    typeChecker;
    subNodeParser;
    expose;
    jsDoc;
    constructor(typeChecker, subNodeParser, expose, jsDoc) {
        this.typeChecker = typeChecker;
        this.subNodeParser = subNodeParser;
        this.expose = expose;
        this.jsDoc = jsDoc;
    }
    supportsNode(node) {
        return this.subNodeParser.supportsNode(node);
    }
    createType(node, context, reference) {
        const baseType = this.subNodeParser.createType(node, context, reference);
        if (!this.isExportNode(node)) {
            return baseType;
        }
        return new DefinitionType(this.getDefinitionName(node, context), baseType);
    }
    isExportNode(node) {
        if (this.expose === "all") {
            return node.kind !== ts.SyntaxKind.TypeLiteral;
        }
        else if (this.expose === "none") {
            return false;
        }
        else if (this.jsDoc !== "none" && hasJsDocTag(node, "internal")) {
            return false;
        }
        const localSymbol = node.localSymbol;
        return localSymbol ? "exportSymbol" in localSymbol : false;
    }
    getDefinitionName(node, context) {
        const symbol = symbolAtNode(node);
        const fullName = this.typeChecker.getFullyQualifiedName(symbol).replace(/^".*"\./, "");
        const argumentIds = context.getArguments().map((arg) => arg?.getName());
        return argumentIds.length ? `${fullName}<${argumentIds.join(",")}>` : fullName;
    }
}
