// ts-json-schema-generator/cjs/factory/program.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.createProgram = void 0;
const tslib_1 = require("tslib");
const glob = tslib_1.__importStar(require("glob"));
const path = tslib_1.__importStar(require("path"));
const typescript_1 = tslib_1.__importDefault(require("typescript"));
const normalize_path_1 = tslib_1.__importDefault(require("normalize-path"));
const DiagnosticError_js_1 = require("../src/Error/DiagnosticError.js");
const LogicError_js_1 = require("../src/Error/LogicError.js");
const NoRootNamesError_js_1 = require("../src/Error/NoRootNamesError.js");
const NoTSConfigError_js_1 = require("../src/Error/NoTSConfigError.js");
function loadTsConfigFile(configFile) {
    const raw = typescript_1.default.sys.readFile(configFile);
    if (raw) {
        const config = typescript_1.default.parseConfigFileTextToJson(configFile, raw);
        if (config.error) {
            throw new DiagnosticError_js_1.DiagnosticError([config.error]);
        }
        else if (!config.config) {
            throw new LogicError_js_1.LogicError(`Invalid parsed config file "${configFile}"`);
        }
        const parseResult = typescript_1.default.parseJsonConfigFileContent(config.config, typescript_1.default.sys, path.resolve(path.dirname(configFile)), {}, configFile);
        parseResult.options.noEmit = true;
        delete parseResult.options.out;
        delete parseResult.options.outDir;
        delete parseResult.options.outFile;
        delete parseResult.options.declaration;
        delete parseResult.options.declarationDir;
        delete parseResult.options.declarationMap;
        return parseResult;
    }
    else {
        throw new NoTSConfigError_js_1.NoTSConfigError();
    }
}
function getTsConfig(config) {
    if (config.tsconfig) {
        return loadTsConfigFile(config.tsconfig);
    }
    return {
        fileNames: [],
        options: {
            noEmit: true,
            emitDecoratorMetadata: true,
            experimentalDecorators: true,
            target: typescript_1.default.ScriptTarget.ES5,
            module: typescript_1.default.ModuleKind.CommonJS,
            strictNullChecks: false,
        },
    };
}
function createProgram(config) {
    const rootNamesFromPath = config.path
        ? glob.sync((0, normalize_path_1.default)(path.resolve(config.path))).map((rootName) => (0, normalize_path_1.default)(rootName))
        : [];
    const tsconfig = getTsConfig(config);
    const rootNames = rootNamesFromPath.length ? rootNamesFromPath : tsconfig.fileNames;
    if (!rootNames.length) {
        throw new NoRootNamesError_js_1.NoRootNamesError();
    }
    const program = typescript_1.default.createProgram(rootNames, tsconfig.options);
    if (!config.skipTypeCheck) {
        const diagnostics = typescript_1.default.getPreEmitDiagnostics(program);
        if (diagnostics.length) {
            throw new DiagnosticError_js_1.DiagnosticError(diagnostics);
        }
    }
    return program;
}
exports.createProgram = createProgram;
