// ts-json-schema-generator/dist/src/Type/ObjectType.js
import { BaseType } from "./BaseType.js";
import { strip } from "../Utils/String.js";
export class ObjectProperty {
    name;
    type;
    required;
    constructor(name, type, required) {
        this.name = name;
        this.type = type;
        this.required = required;
    }
    getName() {
        return strip(this.name);
    }
    getType() {
        return this.type;
    }
    isRequired() {
        return this.required;
    }
}
export class ObjectType extends BaseType {
    id;
    baseTypes;
    properties;
    additionalProperties;
    nonPrimitive;
    constructor(id, baseTypes, properties, additionalProperties, nonPrimitive = false) {
        super();
        this.id = id;
        this.baseTypes = baseTypes;
        this.properties = properties;
        this.additionalProperties = additionalProperties;
        this.nonPrimitive = nonPrimitive;
    }
    getId() {
        return this.id;
    }
    getBaseTypes() {
        return this.baseTypes;
    }
    getProperties() {
        return this.properties;
    }
    getAdditionalProperties() {
        return this.additionalProperties;
    }
    getNonPrimitive() {
        return this.nonPrimitive;
    }
}
