// ts-json-schema-generator/cjs/src/TypeFormatter/PrimitiveUnionTypeFormatter.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.PrimitiveUnionTypeFormatter = void 0;
const LogicError_js_1 = require("../Error/LogicError.js");
const BooleanType_js_1 = require("../Type/BooleanType.js");
const NullType_js_1 = require("../Type/NullType.js");
const NumberType_js_1 = require("../Type/NumberType.js");
const PrimitiveType_js_1 = require("../Type/PrimitiveType.js");
const StringType_js_1 = require("../Type/StringType.js");
const UnionType_js_1 = require("../Type/UnionType.js");
const uniqueArray_js_1 = require("../Utils/uniqueArray.js");
class PrimitiveUnionTypeFormatter {
    supportsType(type) {
        return type instanceof UnionType_js_1.UnionType && type.getTypes().length > 0 && this.isPrimitiveUnion(type);
    }
    getDefinition(type) {
        return {
            type: (0, uniqueArray_js_1.uniqueArray)(type.getTypes().map((item) => this.getPrimitiveType(item))),
        };
    }
    getChildren(type) {
        return [];
    }
    isPrimitiveUnion(type) {
        return type.getTypes().every((item) => item instanceof PrimitiveType_js_1.PrimitiveType);
    }
    getPrimitiveType(item) {
        if (item instanceof StringType_js_1.StringType) {
            return "string";
        }
        else if (item instanceof NumberType_js_1.NumberType) {
            return "number";
        }
        else if (item instanceof BooleanType_js_1.BooleanType) {
            return "boolean";
        }
        else if (item instanceof NullType_js_1.NullType) {
            return "null";
        }
        throw new LogicError_js_1.LogicError("Unexpected code branch");
    }
}
exports.PrimitiveUnionTypeFormatter = PrimitiveUnionTypeFormatter;
