// ts-json-schema-generator/dist/src/TypeFormatter/LiteralUnionTypeFormatter.js
import { LiteralType } from "../Type/LiteralType.js";
import { NullType } from "../Type/NullType.js";
import { StringType } from "../Type/StringType.js";
import { UnionType } from "../Type/UnionType.js";
import { typeName } from "../Utils/typeName.js";
import { uniqueArray } from "../Utils/uniqueArray.js";
export class LiteralUnionTypeFormatter {
    supportsType(type) {
        return type instanceof UnionType && type.getTypes().length > 0 && isLiteralUnion(type);
    }
    getDefinition(type) {
        let hasString = false;
        let preserveLiterals = false;
        let allStrings = true;
        let hasNull = false;
        const literals = type.getFlattenedTypes();
        const types = literals.filter((t) => {
            if (t instanceof StringType) {
                hasString = true;
                preserveLiterals = preserveLiterals || t.getPreserveLiterals();
                return false;
            }
            else if (t instanceof NullType) {
                hasNull = true;
                return true;
            }
            else if (t instanceof LiteralType && !t.isString()) {
                allStrings = false;
            }
            return true;
        });
        if (allStrings && hasString && !preserveLiterals) {
            return {
                type: hasNull ? ["string", "null"] : "string",
            };
        }
        const values = uniqueArray(types.map(getLiteralValue));
        const typeNames = uniqueArray(types.map(getLiteralType));
        const ret = {
            type: typeNames.length === 1 ? typeNames[0] : typeNames,
            enum: values,
        };
        if (preserveLiterals) {
            return {
                anyOf: [
                    {
                        type: "string",
                    },
                    ret,
                ],
            };
        }
        return ret;
    }
    getChildren(type) {
        return [];
    }
}
export function isLiteralUnion(type) {
    return type
        .getFlattenedTypes()
        .every((item) => item instanceof LiteralType || item instanceof NullType || item instanceof StringType);
}
function getLiteralValue(value) {
    return value instanceof LiteralType ? value.getValue() : null;
}
function getLiteralType(value) {
    return value instanceof LiteralType ? typeName(value.getValue()) : "null";
}
