// ts-json-schema-generator/dist/src/TypeFormatter/TupleTypeFormatter.js
import { ArrayType } from "../Type/ArrayType.js";
import { OptionalType } from "../Type/OptionalType.js";
import { RestType } from "../Type/RestType.js";
import { TupleType } from "../Type/TupleType.js";
import { notNever } from "../Utils/notNever.js";
import { uniqueArray } from "../Utils/uniqueArray.js";
function uniformRestType(type, check_type) {
    const inner = type.getType();
    return ((inner instanceof ArrayType && inner.getItem().getId() === check_type.getId()) ||
        (inner instanceof TupleType &&
            inner.getTypes().every((tuple_type) => {
                if (tuple_type instanceof RestType) {
                    return uniformRestType(tuple_type, check_type);
                }
                else {
                    return tuple_type?.getId() === check_type.getId();
                }
            })));
}
export class TupleTypeFormatter {
    childTypeFormatter;
    constructor(childTypeFormatter) {
        this.childTypeFormatter = childTypeFormatter;
    }
    supportsType(type) {
        return type instanceof TupleType;
    }
    getDefinition(type) {
        const subTypes = type.getTypes().filter(notNever);
        const requiredElements = subTypes.filter((t) => !(t instanceof OptionalType) && !(t instanceof RestType));
        const optionalElements = subTypes.filter((t) => t instanceof OptionalType);
        const restType = subTypes.find((t) => t instanceof RestType);
        const firstItemType = requiredElements.length > 0 ? requiredElements[0] : optionalElements[0]?.getType();
        const isUniformArray = firstItemType &&
            requiredElements.every((item) => item.getId() === firstItemType.getId()) &&
            optionalElements.every((item) => item.getType().getId() === firstItemType.getId()) &&
            (!restType || uniformRestType(restType, firstItemType));
        if (isUniformArray) {
            return {
                type: "array",
                items: this.childTypeFormatter.getDefinition(firstItemType),
                minItems: requiredElements.length,
                ...(restType ? {} : { maxItems: requiredElements.length + optionalElements.length }),
            };
        }
        const requiredDefinitions = requiredElements.map((item) => this.childTypeFormatter.getDefinition(item));
        const optionalDefinitions = optionalElements.map((item) => this.childTypeFormatter.getDefinition(item));
        const itemsTotal = requiredDefinitions.length + optionalDefinitions.length;
        const additionalItems = restType ? this.childTypeFormatter.getDefinition(restType).items : undefined;
        return {
            type: "array",
            minItems: requiredDefinitions.length,
            ...(itemsTotal ? { items: requiredDefinitions.concat(optionalDefinitions) } : {}),
            ...(!itemsTotal && additionalItems ? { items: additionalItems } : {}),
            ...(!itemsTotal && !additionalItems ? { maxItems: 0 } : {}),
            ...(additionalItems && !Array.isArray(additionalItems) && itemsTotal
                ? { additionalItems: additionalItems }
                : {}),
            ...(!additionalItems && itemsTotal ? { maxItems: itemsTotal } : {}),
        };
    }
    getChildren(type) {
        return uniqueArray(type
            .getTypes()
            .filter(notNever)
            .reduce((result, item) => [...result, ...this.childTypeFormatter.getChildren(item)], []));
    }
}
