// ts-json-schema-generator/dist/src/TypeFormatter/UnionTypeFormatter.js
import { LiteralType } from "../Type/LiteralType.js";
import { NeverType } from "../Type/NeverType.js";
import { UnionType } from "../Type/UnionType.js";
import { derefType } from "../Utils/derefType.js";
import { getTypeByKey } from "../Utils/typeKeys.js";
import { uniqueArray } from "../Utils/uniqueArray.js";
export class UnionTypeFormatter {
    childTypeFormatter;
    discriminatorType;
    constructor(childTypeFormatter, discriminatorType) {
        this.childTypeFormatter = childTypeFormatter;
        this.discriminatorType = discriminatorType;
    }
    supportsType(type) {
        return type instanceof UnionType;
    }
    getTypeDefinitions(type) {
        return type
            .getTypes()
            .filter((item) => !(derefType(item) instanceof NeverType))
            .map((item) => this.childTypeFormatter.getDefinition(item));
    }
    getJsonSchemaDiscriminatorDefinition(type) {
        const definitions = this.getTypeDefinitions(type);
        const discriminator = type.getDiscriminator();
        if (!discriminator)
            throw new Error("discriminator is undefined");
        const kindTypes = type
            .getTypes()
            .filter((item) => !(derefType(item) instanceof NeverType))
            .map((item) => getTypeByKey(item, new LiteralType(discriminator)));
        const undefinedIndex = kindTypes.findIndex((item) => item === undefined);
        if (undefinedIndex != -1) {
            throw new Error(`Cannot find discriminator keyword "${discriminator}" in type ${type.getTypes()[undefinedIndex].getName()}.`);
        }
        const kindDefinitions = kindTypes.map((item) => this.childTypeFormatter.getDefinition(item));
        const allOf = [];
        for (let i = 0; i < definitions.length; i++) {
            allOf.push({
                if: {
                    properties: { [discriminator]: kindDefinitions[i] },
                },
                then: definitions[i],
            });
        }
        const kindValues = kindDefinitions
            .flatMap((item) => item.const ?? item.enum)
            .filter((item) => item !== undefined);
        const duplicates = kindValues.filter((item, index) => kindValues.indexOf(item) !== index);
        if (duplicates.length > 0) {
            throw new Error(`Duplicate discriminator values: ${duplicates.join(", ")} in type ${JSON.stringify(type.getName())}.`);
        }
        const properties = {
            [discriminator]: {
                enum: kindValues,
            },
        };
        return { type: "object", properties, required: [discriminator], allOf };
    }
    getOpenApiDiscriminatorDefinition(type) {
        const oneOf = this.getTypeDefinitions(type);
        const discriminator = type.getDiscriminator();
        if (!discriminator)
            throw new Error("discriminator is undefined");
        return {
            type: "object",
            discriminator: { propertyName: discriminator },
            required: [discriminator],
            oneOf,
        };
    }
    getDefinition(type) {
        const discriminator = type.getDiscriminator();
        if (discriminator !== undefined) {
            if (this.discriminatorType === "open-api")
                return this.getOpenApiDiscriminatorDefinition(type);
            return this.getJsonSchemaDiscriminatorDefinition(type);
        }
        const definitions = this.getTypeDefinitions(type);
        const flattenedDefinitions = [];
        for (const def of definitions) {
            const keys = Object.keys(def);
            if (keys.length === 1 && keys[0] === "anyOf") {
                flattenedDefinitions.push(...def.anyOf);
            }
            else {
                flattenedDefinitions.push(def);
            }
        }
        return flattenedDefinitions.length > 1
            ? {
                anyOf: flattenedDefinitions,
            }
            : flattenedDefinitions[0];
    }
    getChildren(type) {
        return uniqueArray(type
            .getTypes()
            .reduce((result, item) => [...result, ...this.childTypeFormatter.getChildren(item)], []));
    }
}
